jsPsych subject feedback

gdoubleyew opened the pull request:
Initial implementation of using jsPsych to give subject feedback via a web browser.

s-michelmann commented on web/html/jsPsychFeedback.html:
```diff
@@ -0,0 +1,115 @@
+  var minDim = Math.min(width, height)
+  var brainstate = stateVal * 0.66 * minDim
+    ctx.beginPath();
+    ctx.arc(width/2, height/2, 1*brainstate, 0, 2 * Math.PI);
```
change brainstate to radius

s-michelmann commented on web/html/jsPsychFeedback.html:
```diff
@@ -0,0 +1,115 @@
+  var width = canvas.width
+  var height = canvas.height
+  var minDim = Math.min(width, height)
+  var brainstate = stateVal * 0.66 * minDim
```
update this variable with an eval() command?

the function would be used like this: 

drawFunction  = drawCircle
drawFunction(parameter, stateVal)

parameter would be e.g. the string "width" then, in drawCircle we call eval(parameter + '= '+ string(stateVal)

gdoubleyew commented:
Hi Sebastian,
I addressed these comments in the latest commit, see diff here,
https://github.com/brainiak/rt-cloud/commit/d464c32cd78f5ee60fd0e3c824dfd927669c02a7

The only thing I didn’t do was add a drawImage() function because it wasn’t clear the best way to get the images to the feedback webpage, so I deferred that to another time when it might be needed or more obvious.

Thanks,
Grant

From: s-michelmann ***@***.***>
Reply-To: brainiak/rt-cloud ***@***.***>
Date: Tuesday, September 7, 2021 at 5:31 PM
To: brainiak/rt-cloud ***@***.***>
Cc: Grant Wallace ***@***.***>, Author ***@***.***>
Subject: Re: [brainiak/rt-cloud] jsPsych subject feedback (#62)


@s-michelmann commented on this pull request.

________________________________

In rtCommon/subjectInterface.py<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703779783>:

>          the presentation script to later read and use to provide subject feedback.

         Args:

             runId: experiment specific identifier of the run

             trId: volume number of the dicom within a run

             value: the classification result from processing the dicom image for this TR

+            msTimeDelay: time in milliseconds to wait before presenting the feedback stimulus

maybe rename?
onset_time_delay?

________________________________

In rtCommon/subjectInterface.py<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703785778>:

>

     def getResponse(self, runId :int, trId :int):

         """

         Retrieve the subject response, used by the classification script.

         """

         print(f'SubjectInterface: getResponse: run {runId}, tr {trId}')

         pass

+

+    def getResponses(self):

+        """

+        Retrieve all subject responses since the last time this call was made

+        """

+        print(f'SubjectInterface: getResponses')

+        pass

maybe add a comment when those get called

________________________________

In rtCommon/subjectInterface.py<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703790963>:

>

     def getResponse(self, runId :int, trId :int):

         """

         Retrieve the subject response, used by the classification script.

         """

         print(f'SubjectInterface: getResponse: run {runId}, tr {trId}')

         pass

+

+    def getResponses(self):

+        """

+        Retrieve all subject responses since the last time this call was made

+        """

+        print(f'SubjectInterface: getResponses')

+        pass

+

+    def dequeueResult(self, block :bool=False, timeout :int=None) -> float:

add a comment on how this works too. (if in local mode, what is being returned?)

________________________________

In tests/test_subjectInterface.py<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703795214>:

> @@ -37,10 +37,12 @@ def runSubjectFeedbackTest(isRemote):

     runId = 3

     for trId in range(1, 10):

test for float?
maybe random integers/numbers?
what breaks it? error handling? what is the dumbest entry that a user can make? -99?

________________________________

In web/html/jsPsychFeedback.html<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703796813>:

> +<html>

+

+<head>

+  <meta charset="UTF-8">

+  <title>jsPsych Subject Feedback</title>

+  <script src="jspsych/jspsych.js"></script>

+  <link href="jspsych/css/jspsych.css" rel="stylesheet" type="text/css">

+  <script src="src/jspsych-plugin-realtime-response.js"></script>

+  <script src="src/jsPsychWebSocket.js"></script>

+</head>

+

+<body></body>

+

+<script>

+// connect to the project server to receive events

+createWebSocket();

try catch?
maybe think about error handling if this fails. Refresh the browser n-times, then exit the program in a clean fashion

________________________________

In web/src/jspsych-plugin-realtime-response.js<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703802747>:

> +      keyboardListener = jsPsych.pluginAPI.getKeyboardResponse({

+        callback_function: after_response,

+        valid_responses: trial.choices,

+        rt_method: 'performance',

+        persist: false,

+        allow_held_key: false,

+      });

+    }

+

+    // function to wrap up when the event has arrived (not finished if it's too early)

+    function event_handler(e){

+      // remove the listener first!

+      // TODO - maybe don't remove listener? - how would we cancel a previous set timeout?

+      document.removeEventListener('rtEvent', event_handler);

+

+      var msTimeDelay = 0;

maybe comment to explain

________________________________

In web/src/jspsych-plugin-realtime-response.js<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703803088>:

> +

+    // function to wrap up when the event has arrived (not finished if it's too early)

+    function event_handler(e){

+      // remove the listener first!

+      // TODO - maybe don't remove listener? - how would we cancel a previous set timeout?

+      document.removeEventListener('rtEvent', event_handler);

+

+      var msTimeDelay = 0;

+      if ("detail" in e) {

+        if ("msTimeDelay" in e.detail) {

+          msTimeDelay = e.detail.msTimeDelay;

+        }

+      }

+

+      // store when the rtEvent event happened

+      rtEvent_time = performance.now() - start_time;

rename to socket_eventTime

________________________________

In web/src/jspsych-plugin-realtime-response.js<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703810730>:

> +

+      var msTimeDelay = 0;

+      if ("detail" in e) {

+        if ("msTimeDelay" in e.detail) {

+          msTimeDelay = e.detail.msTimeDelay;

+        }

+      }

+

+      // store when the rtEvent event happened

+      rtEvent_time = performance.now() - start_time;

+

+      if (msTimeDelay > 0) {

+        // setTimeout input is time in miliseconds

+        jsPsych.pluginAPI.setTimeout(end_trial, msTimeDelay)

+      } else if ((trial.trial_duration !== null) && (rtEvent_time < trial.trial_duration)){

+        // if the setResult was early we want to sleep for a bit

this is redundant now, but the msTimeDelay needs to be fail safe and clear (how to set it etc.)

Also, maybe try to log the deviation of current onset from expected TR time. e.g.,
... current presentation-to-presentation-lag = 2.8s,
also: the TR-header (that has the onset) and the rtEvent_time difference should remain constant over the exp
(this is a job for the analysis computer to compare, so send absolute stimulus onset time back to analysis computer)

________________________________

In web/src/jspsych-plugin-realtime-response.js<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703822694>:

> +        jsPsych.pluginAPI.setTimeout(end_trial, (trial.trial_duration-rtEvent_time))

+      } else{

+        //otherwise we just end the trial now

+        end_trial()

+      }

+    }

+

+    // function to end trial when it is time

+    function end_trial() {

+      if (keyboardListener != null) {

+        jsPsych.pluginAPI.cancelKeyboardResponse(keyboardListener)

+      }

+

+      // gather the data to store for the trial

+      var trial_data = {

+        "rtEvent_time":  rtEvent_time,

also store: stimulus-to-stimulus, which is performance.now()-start_time. (may not be needed anymore, see second point)
also store: absolute stimulus onset time (performance.now()) for the computation of difference TR-header-to-Stim on the analysis computer

________________________________

In web/src/jspsych-plugin-realtime-response.js<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703824497>:

> +

+  var plugin = {};

+

+  // TODO (Improvement) ideally the plugin should have a parameter that is reserved for an element that

+  //   the listener should be attached to! (e.g. the websocket)

+  plugin.info = {

+    name: 'brain-realtime-response',

+    description: '',

+    parameters: {

+      trial_duration: {

+        type: jsPsych.plugins.parameterType.INT,

+        pretty_name: 'Trial duration',

+        default: null,

+        description: 'The minimal duration of the trial in milliseconds.'

+      },

+      canvas_size: {

maybe not necessary any longer

________________________________

In web/html/jsPsychFeedback.html<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703825281>:

> +  <script src="jspsych/jspsych.js"></script>

+  <link href="jspsych/css/jspsych.css" rel="stylesheet" type="text/css">

+  <script src="src/jspsych-plugin-realtime-response.js"></script>

+  <script src="src/jsPsychWebSocket.js"></script>

+</head>

+

+<body></body>

+

+<script>

+// connect to the project server to receive events

+createWebSocket();

+

+var timeline = []

+

+// this function draws a circle with a brainstate radius

+function drawCircle(canvas, stateVal){

maybe have a few function examples, e.g., drawRectangle()., drawImage() Then you can change what the drawFunction points to.

drawFunction = drawCircle

________________________________

In web/html/jsPsychFeedback.html<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703839756>:

> +

+<script>

+// connect to the project server to receive events

+createWebSocket();

+

+var timeline = []

+

+// this function draws a circle with a brainstate radius

+function drawCircle(canvas, stateVal){

+  var ctx = canvas.getContext('2d');

+  var width = canvas.width

+  var height = canvas.height

+  var minDim = Math.min(width, height)

+  var brainstate = stateVal * 0.66 * minDim

+    ctx.beginPath();

+    ctx.arc(width/2, height/2, 1*brainstate, 0, 2 * Math.PI);

change brainstate to radius

________________________________

In web/html/jsPsychFeedback.html<https://github.com/brainiak/rt-cloud/pull/62#discussion_r703843584>:

> +

+<body></body>

+

+<script>

+// connect to the project server to receive events

+createWebSocket();

+

+var timeline = []

+

+// this function draws a circle with a brainstate radius

+function drawCircle(canvas, stateVal){

+  var ctx = canvas.getContext('2d');

+  var width = canvas.width

+  var height = canvas.height

+  var minDim = Math.min(width, height)

+  var brainstate = stateVal * 0.66 * minDim

update this variable with an eval() command?

the function would be used like this:

drawFunction = drawCircle
drawFunction(parameter, stateVal)

parameter would be e.g. the string "width" then, in drawCircle we call eval(parameter + '= '+ string(stateVal)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/brainiak/rt-cloud/pull/62#pullrequestreview-748333251>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AH7ZDPJFD3TUY5NXI2XTQ4DUA2AERANCNFSM46BQJ65A>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
